File: experiments/adrianco/experiment-16-qwen3coder/bookshop-128k/runs/agent=qwen3-coder-local_language=go/rep1/handler.go
package main

import (
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/gorilla/mux"
)

// Handler handles HTTP requests for the book API
type Handler struct {
	db BookStore
}

// createBook handles POST /books
func (h *Handler) createBook(w http.ResponseWriter, r *http.Request) {
	var book Book

	// Parse the request body
	if err := json.NewDecoder(r.Body).Decode(&book); err != nil {
		http.Error(w, "Invalid JSON", http.StatusBadRequest)
		return
	}

	// Validate required fields
	if book.Title == "" {
		http.Error(w, "Title is required", http.StatusBadRequest)
		return
	}
	if book.Author == "" {
		http.Error(w, "Author is required", http.StatusBadRequest)
		return
	}

	// Save the book
	if err := h.db.CreateBook(&book); err != nil {
		http.Error(w, "Failed to create book", http.StatusInternalServerError)
		return
	}

	// Return the created book with ID
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(book)
}

// getBooks handles GET /books with optional author filter
func (h *Handler) getBooks(w http.ResponseWriter, r *http.Request) {
	// Get the author query parameter
	author := r.URL.Query().Get("author")

	// Get books from the database
	books, err := h.db.GetBooks(author)
	if err != nil {
		http.Error(w, "Failed to retrieve books", http.StatusInternalServerError)
		return
	}

	// Return the books
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(books)
}

// getBook handles GET /books/{id}
func (h *Handler) getBook(w http.ResponseWriter, r *http.Request) {
	// Get the book ID from the URL
	vars := mux.Vars(r)
	id, err := strconv.Atoi(vars["id"])
	if err != nil {
		http.Error(w, "Invalid book ID", http.StatusBadRequest)
		return
	}

	// Get the book from the database
	book, err := h.db.GetBook(id)
	if err != nil {
		if err.Error() == "sql: no rows in result set" {
			http.Error(w, "Book not found", http.StatusNotFound)
		} else {
			http.Error(w, "Failed to retrieve book", http.StatusInternalServerError)
		}
		return
	}

	// Return the book
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(book)
}

// updateBook handles PUT /books/{id}
func (h *Handler) updateBook(w http.ResponseWriter, r *http.Request) {
	// Get the book ID from the URL
	vars := mux.Vars(r)
	id, err := strconv.Atoi(vars["id"])
	if err != nil {
		http.Error(w, "Invalid book ID", http.StatusBadRequest)
		return
	}

	// Parse the request body
	var book Book
	if err := json.NewDecoder(r.Body).Decode(&book); err != nil {
		http.Error(w, "Invalid JSON", http.StatusBadRequest)
		return
	}

	// Validate required fields
	if book.Title == "" {
		http.Error(w, "Title is required", http.StatusBadRequest)
		return
	}
	if book.Author == "" {
		http.Error(w, "Author is required", http.StatusBadRequest)
		return
	}

	// Update the book in the database
	if err := h.db.UpdateBook(id, &book); err != nil {
		if err.Error() == "sql: no rows in result set" {
			http.Error(w, "Book not found", http.StatusNotFound)
		} else {
			http.Error(w, "Failed to update book", http.StatusInternalServerError)
		}
		return
	}

	// Return the updated book
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(book)
}

// deleteBook handles DELETE /books/{id}
func (h *Handler) deleteBook(w http.ResponseWriter, r *http.Request) {
	// Get the book ID from the URL
	vars := mux.Vars(r)
	id, err := strconv.Atoi(vars["id"])
	if err != nil {
		http.Error(w, "Invalid book ID", http.StatusBadRequest)
		return
	}

	// Delete the book from the database
	if err := h.db.DeleteBook(id); err != nil {
		if err.Error() == "sql: no rows in result set" {
			http.Error(w, "Book not found", http.StatusNotFound)
		} else {
			http.Error(w, "Failed to delete book", http.StatusInternalServerError)
		}
		return
	}

	// Return success response
	w.WriteHeader(http.StatusNoContent)
}

// healthCheck handles GET /health
func (h *Handler) healthCheck(w http.ResponseWriter, r *http.Request) {
	if err := h.db.HealthCheck(); err != nil {
		http.Error(w, "Database health check failed", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(map[string]string{"status": "healthy"})
}

File: experiments/adrianco/experiment-6/runs/language=clojure_model=claude-opus-4-7_tooling=beads/rep2/src/books/handler.clj
(ns books.handler
  (:require [compojure.core :refer [defroutes GET POST PUT DELETE]]
            [compojure.route :as route]
            [ring.middleware.json :refer [wrap-json-body wrap-json-response]]
            [ring.middleware.params :refer [wrap-params]]
            [ring.util.response :as resp]
            [books.db :as db]))

(defn- json-response
  ([status body]
   (-> (resp/response body)
       (resp/status status)
       (resp/header "Content-Type" "application/json"))))

(defn- parse-id [s]
  (try (Long/parseLong s) (catch Exception _ nil)))

(defn- validate-book [{:keys [title author year]}]
  (cond
    (or (nil? title) (and (string? title) (clojure.string/blank? title)))
    "title is required"

    (or (nil? author) (and (string? author) (clojure.string/blank? author)))
    "author is required"

    (and (some? year) (not (integer? year)))
    "year must be an integer"

    :else nil))

(defn create-book-handler [ds]
  (fn [req]
    (let [body (:body req)
          err (validate-book body)]
      (if err
        (json-response 400 {:error err})
        (let [book (db/create-book ds body)]
          (json-response 201 book))))))

(defn list-books-handler [ds]
  (fn [req]
    (let [author (get-in req [:query-params "author"])
          books (db/list-books ds author)]
      (json-response 200 books))))

(defn get-book-handler [ds]
  (fn [req]
    (let [id (parse-id (get-in req [:params :id]))]
      (if-let [book (and id (db/get-book ds id))]
        (json-response 200 book)
        (json-response 404 {:error "Book not found"})))))

(defn update-book-handler [ds]
  (fn [req]
    (let [id (parse-id (get-in req [:params :id]))
          body (:body req)
          err (validate-book body)]
      (cond
        (nil? id) (json-response 404 {:error "Book not found"})
        err (json-response 400 {:error err})
        :else
        (if-let [book (db/update-book ds id body)]
          (json-response 200 book)
          (json-response 404 {:error "Book not found"}))))))

(defn delete-book-handler [ds]
  (fn [req]
    (let [id (parse-id (get-in req [:params :id]))]
      (if (and id (db/delete-book ds id))
        (json-response 204 nil)
        (json-response 404 {:error "Book not found"})))))

(defn health-handler [_req]
  (json-response 200 {:status "ok"}))

(defn make-app [ds]
  (let [routes (compojure.core/routes
                (GET "/health" [] health-handler)
                (POST "/books" [] (create-book-handler ds))
                (GET "/books" [] (list-books-handler ds))
                (GET "/books/:id" [] (get-book-handler ds))
                (PUT "/books/:id" [] (update-book-handler ds))
                (DELETE "/books/:id" [] (delete-book-handler ds))
                (route/not-found {:error "Not found"}))]
    (-> routes
        (wrap-json-body {:keywords? true})
        wrap-json-response
        wrap-params)))
